internal/app/db.go

```diff
@@ -0,0 +1,331 @@
+		}
+	}
+
+	query := dbm.queryFromWAL(tx)
+	slog.Info("replaying", "query", query)
+
+	_, err := dbm.db.ExecContext(ctx, query)
+	if err != nil {
+		return fmt.Errorf("cannot replay WAL record %s", err)
+	}
+
+	return nil
+}
+
+func (dbm *DBManager) genCreateQuery() (string, error) {
+	var cols, pks string
+	for i, column := range dbm.cols {
+		col := fmt.Sprintf("%s %s", column.Name, column.Typ)
+		if !column.IsNull {
+			col = fmt.Sprintf("%s NOT NULL", col)
+		}
+		if i == 0 {
+			cols = col
+			if column.IsPrimary {
+				pks = column.Name
+			}
+		} else {
+			cols = fmt.Sprintf("%s,%s", cols, col)
+			if column.IsPrimary {
+				pks = fmt.Sprintf("%s,%s", pks, column.Name)
+			}
+		}
+	}
+
+	if pks != "" {
+		cols = fmt.Sprintf("%s,PRIMARY KEY (%s)", cols, pks)
+	}
+
+	if cols == "" {
+		return "", errors.New("schema must have at least one column")
+	}
+
+	stmt := fmt.Sprintf(
+		"CREATE TABLE IF NOT EXISTS %s (%s)",
+		dbm.table, cols)
+	return stmt, nil
+}
+
+// Export exports the current db to a parquet file at the given path.
+func (dbm *DBManager) Export(ctx context.Context, exportPath string) error {
+	var err error
+	db := dbm.db
+	// db is nil before replication starts.
+	// In that case, we open all existing db files
+	// and upload them.
+	if db == nil {
+		// convert the export path to a db path:
+		// <ts>.db.parquet -> <ts>.db
+		dbPath := strings.ReplaceAll(exportPath, ".parquet", "")
+		db, err = sql.Open("duckdb", dbPath)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err := db.Close(); err != nil {
+				fmt.Printf("cannot close db %v \n", err)
+			}
+		}()
+		slog.Info("backing up db", "at", exportPath)
+	} else {
+		slog.Info("backing up current db")
+	}
+	_, err = db.ExecContext(ctx,
+		fmt.Sprintf(
+			`INSTALL parquet;
+			 LOAD parquet;
+			 COPY (SELECT * FROM %s) TO '%s' (FORMAT PARQUET)`,
+			dbm.table, exportPath))
+	if err != nil {
+		return fmt.Errorf("cannot export to parquet file: %s", err)
+	}
+
+	return nil
+}
+
+func (dbm *DBManager) cleanup(dbPath string) error {
+	slog.Info("deleting db dump", "at", dbPath)
+	if err := os.Remove(dbPath); err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("cannot delete file: %s", err)
+		}
+	}
+
+	walPath := dbPath + ".wal"
+	slog.Info("deleting db wal", "at", walPath)
+	if err := os.Remove(walPath); err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("cannot delete file: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// UploadAt uploads a db dump at the given path.
+// It returns an error if a dumps cannot be uploaded.
+// It cleans up the db dumps and export files after uploading.
+func (dbm *DBManager) UploadAt(ctx context.Context, exportPath string) error {
+	f, err := os.Open(exportPath)
+	if err != nil {
+		return fmt.Errorf("cannot open file: %s", err)
+	}
+
+	fi, err := f.Stat()
+	if err != nil {
+		return fmt.Errorf("cannot stat file: %s", err)
+	}
+
+	progress := progressbar.DefaultBytes(
```

Comment:
not sure if the progress bar makes sense in a background routine, this is more useful for a user when executing a command

Reply:
can pass in `nil` to remove it but it looks okay to me since you can tell that the "_replication is paused while this file uploads_". 

may we can change it later based on the user feedback